pybop/plotting/plot2d.py

```diff
@@ -141,7 +141,7 @@ def plot2d(
         parameter_log = np.asarray(optim.log["x_best"])
         flat_x = np.concatenate((flat_x, parameter_log[:, 0]))
         flat_y = np.concatenate((flat_y, parameter_log[:, 1]))
-        flat_costs = np.concatenate((flat_costs, optim.log["cost"]))
+        flat_costs = np.concatenate((flat_costs, optim.log["cost_best"]))
```

Comment:
Suggestion for lines 102-144 to reduce number of calls when the gradient is requested and to include all `x`, rather than just `x_best`, in the trace now that the corresponding `cost_best` are available.
```
    if gradient:
        grad_parameter_costs = []

        # Determine the number of gradient outputs from cost.compute
        num_gradients = len(
            cost_call(
                np.asarray([x[0], y[0]] + additional_values),
                calculate_grad=True,
            )[1]
        )

        # Create an array to hold each gradient output
        grads = [np.zeros((len(y), len(x))) for _ in range(num_gradients)]

    # Populate cost matrix
    for i, xi in enumerate(x):
        for j, yj in enumerate(y):
            if not gradient:
                costs[j, i] = cost_call(
                    np.asarray([xi, yj] + additional_values),
                )

            else:
                costs[j, i], (*current_grads,) = cost_call(
                    np.asarray([xi, yj] + additional_values),
                    calculate_grad=True,
                )
                for k, grad_output in enumerate(current_grads):
                    grads[k][j, i] = grad_output

    if gradient:
        # Append the gradient arrays to the grad_parameter_costs list
        grad_parameter_costs.extend(grads)

    if plot_optim and use_optim_log:
        # Flatten the cost matrix and parameter values
        flat_x = np.tile(x, len(y))
        flat_y = np.repeat(y, len(x))
        flat_costs = costs.flatten()

        # Append the optimisation trace to the data
        parameter_log = np.asarray(optim.log["x"])
        flat_x = np.concatenate((flat_x, parameter_log[:, 0]))
        flat_y = np.concatenate((flat_y, parameter_log[:, 1]))
        flat_costs = np.concatenate((flat_costs, optim.log["cost"]))
```

Reply:
Thanks, that's a nice improvement.